### ISSUE TITLE
update_cluster_layout_tests.py::TestUpdateClusterLayout::test_add_new_node_while_add_new_table[after] fails

### ISSUE THREAD

**Issue by gleb-cloudius** [scylladb/scylladb#21371]
The test fails with `E   cassandra.InvalidRequest: Error from server: code=2200 [Invalid query] message="unconfigured table cf1`

Logs attached:
[logs-full.release.000.zip](https://github.com/user-attachments/files/17568867/logs-full.release.000.zip)

**Comment by gleb-cloudius** [comment]
Looks strange. It fails in `create_cf` last line where it tries to select from just created table after waiting for schema agreement. the failure happens at 03:53:43,841 but ` grep "Creating ks1.cf1" node*.log` shows:

```
node1.log:INFO  2024-10-30 03:53:42,232 [shard 0: gms] schema_tables - Creating ks1.cf1 id=8ddd3f60-9672-11ef-be6f-6dd46f498a91 version=8ddd3f61-9672-11ef-be6f-6dd46f498a91
node2.log:INFO  2024-10-30 03:53:42,236 [shard 0: gms] schema_tables - Creating ks1.cf1 id=8ddd3f60-9672-11ef-be6f-6dd46f498a91 version=8ddd3f61-9672-11ef-be6f-6dd46f498a91
node3.log:INFO  2024-10-30 03:53:42,236 [shard 0: gms] schema_tables - Creating ks1.cf1 id=8ddd3f60-9672-11ef-be6f-6dd46f498a91 version=8ddd3f61-9672-11ef-be6f-6dd46f498a91
node4.log:INFO  2024-10-30 03:53:42,236 [shard 0: gms] schema_tables - Creating ks1.cf1 id=8ddd3f60-9672-11ef-be6f-6dd46f498a91 version=8ddd3f61-9672-11ef-be6f-6dd46f498a91
```

So by this point each node created the table already.

**Comment by kbr-scylla** [comment]
> [logs-full.release.000.zip](https://github.com/user-attachments/files/17568867/logs-full.release.000.zip)

posting to prevent wasting more time in the future:
the actual test log is in dtest-gw1.log

**Comment by kbr-scylla** [comment]
Where did you get this code? The Python driver code looks different: https://github.com/scylladb/python-driver/blob/059e10e53c6fc828487896353b46d7f226fec9ac/cassandra/cluster.py#L4148

in particular it skips nodes that driver doesn't know of or it doesn't consider them as UP yet (in `_get_schema_mismatches`):
```python3
            peer = self._cluster.metadata.get_host(endpoint)
            if peer and peer.is_up is not False:
                versions[schema_ver].add(endpoint)
```

in the test log soon before the failure we see
```
03:53:42,223 685     cassandra.cluster              INFO     thread.py           :58   | test_add_new_node_while_add_new_table[after]: New Cassandra host <Host: 127.0.24.4:9042 datacenter1> discovered
```

so it could be that when the schema agreement wait was done, this node was not part of driver's _cluster.metadata yet (i.e. it happened before the "discovery")

**Comment by gleb-cloudius** [comment]
> Where did you get this code? 

It is in the `tools/data.py` in dtest repo. Is this not the one called here?

**Comment by kbr-scylla** [comment]
`create_cf` calls:
```python3
    session.cluster.control_connection.wait_for_schema_agreement(wait_time=120)
```

I don't see a call to the function you posted

**Comment by kbr-scylla** [comment]
So a sequence of events which could explain this failure:
- node 4 joins the cluster. Driver doesn't yet learn about it
- we create `cf`, response is returned after quorum commits, the schema change wasn't applied on node4 yet
- driver waits for schema agreement -- but skips node 4 in the set used for agreement calculation, so it finishes before node4 applies the change
- only now driver learns about node 4
- send query to node 4 using `cf` -- fail, `cf` missing

Unfortunately the communication between driver and cluster does not correctly enforce causality propagation. The heuristics used (e.g. schema agreement wait) are flaky. A potential solution would be to attach version numbers / fencing tokens / however you want to call it, to driver request and responses, and if server receives a request with schema version/token it doesn't know about, it would have to wait (barrier) before handling it. cc @kostja

**Comment by kbr-scylla** [comment]
Other than that, the dtest could be patched to wait for driver to learn about node4 before performing the requests

**Comment by AbhinavKumarJha** [comment]
This issue has been extensively explored and solutions have been discussed in this document: [doc link](https://docs.google.com/document/d/13BTHL42vZBOuIwustZ_GFj_lfVcp77MPwm2it3uMGu0/edit?tab=t.0#heading=h.wmb6zzun2mva)

Since this driver support feature is low priority, leaving this issue as P3, for when we get back to this.

cc: @emdroid @gleb-cloudius

**Comment by calebxyz** [comment]
@emdroid please check if its a duplicate of #25530

**Comment by calebxyz** [comment]
@AbhinavKumarJha if your still working on this p[lease update

**Comment by AbhinavKumarJha** [comment]
>@AbhinavKumarJha if your still working on this p[lease update

As per the last discussion with team, it was deemed a low priority feature and driver team support is also needed for the same. 

@emdroid can we move it to `someday` tag?

**Comment by emdroid** [comment]
Yes, I think we can. Also, the issue didn't reproduce for more than a year, so I think we can remove the "CI stability" tag.
